src/app/reporting/SynchronizedReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/reporting/SynchronizedReportSchedulerImpl.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+
+using Seconds16       = System::Clock::Seconds16;
+using Milliseconds32  = System::Clock::Milliseconds32;
+using Timeout         = System::Clock::Timeout;
+using Timestamp       = System::Clock::Timestamp;
+using ReadHandlerNode = ReportScheduler::ReadHandlerNode;
+
+/// @brief When a ReadHandler becomes reportable, recaculate the earliest possible interval before scheduling an engine run and
+/// reschedule the report
+void SynchronizedReportSchedulerImpl::OnBecameReportable(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturn(nullptr != node);
+    Milliseconds32 newTimeout;
+    ReturnOnFailure(CalculateNextReportTimeout(newTimeout, node));
+    ScheduleReport(newTimeout, node);
+}
+
+void SynchronizedReportSchedulerImpl::OnSubscriptionAction(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturn(nullptr != node);
+
+    // Update the node's timestamps
+    node->SetIntervalTimeStamps(aReadHandler);
+
+    Milliseconds32 newTimeout;
+    ReturnOnFailure(CalculateNextReportTimeout(newTimeout, node));
```

Comment:
So what happens here if ` CalculateNextReportTimeout` fails? They will be no report of this SubscriptionsAction and we just silently return? Shouldn't the error be returned and dealt with by the caller? Same with `OnBecameReportable`